lib/dal/src/workspace_snapshot/migrator.rs

```diff
@@ -77,95 +83,105 @@ impl SnapshotGraphMigrator {
+                let mut ctx_after_migration =
+                    ctx.clone_with_new_visibility(Visibility::from(change_set.id));
+                // TODO make sure that when we clone with a changeset id we also set changeset
+                // (or there's no clone anymore and we always change it via following method)
+                ctx_after_migration.set_change_set(change_set.clone())?;
+
+                let snapshot_address = change_set.workspace_snapshot_address;
+
+                let new_snapshot_address = match migration_map.get(&snapshot_address) {
+                    Some(cached_addr) => *cached_addr,
+                    None => match self
+                        .migrate_snapshot(&ctx_after_migration, snapshot_address)
+                        .await
+                    {
+                        Ok(addr) => addr,
+                        Err(err) => {
+                            let err_string = err.to_string();
+                            if err_string.contains("missing from store for node")
+                                || err_string
+                                    .contains("workspace snapshot graph missing at address")
+                            {
+                                error!(
+                                    error = ?err,
+                                    "Migration error: {err_string}, marking change set {} for workspace {:?} as failed",
+                                    change_set.id, change_set.workspace_id
+                                );
+
+                                change_set
+                                    .update_status(ctx, ChangeSetStatus::Failed)
+                                    .await?;
+                                continue;
+                            } else {
+                                return Err(err)?;
+                            }
+                        }
+                    },
+                };
+
+                let migrated_snapshot =
+                    SplitSnapshot::find(&ctx_after_migration, new_snapshot_address).await?;
+                ctx_after_migration.set_workspace_split_snapshot(migrated_snapshot);
+
+                change_set
+                    .update_pointer(&ctx_after_migration, new_snapshot_address)
+                    .await?;
+
+                migration_map.insert(snapshot_address, new_snapshot_address);
+                change_set_count += 1;
+            }
+
+            workspace
+                .set_snapshot_versions(
+                    ctx,
+                    SnapshotVersion::Split(SuperGraphVersionDiscriminants::V1),
+                    Some(SubGraphVersionDiscriminants::V1),
+                )
                 .await?;
+            workspace_count += 1;
         }
 
-        info!("Migration finished, marking all workspaces as migrated to latest version");
-
-        Workspace::set_snapshot_version_for_all_workspaces(
-            ctx,
-            WorkspaceSnapshotGraph::current_discriminant(),
```

Comment:
Oh man, did we fail to set the version on some workspaces before if it failed halfway? Or was a transaction protecting us here?